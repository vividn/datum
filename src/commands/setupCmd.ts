import { Argv } from "yargs";
import { BaseDatumArgs } from "../input/baseYargs";
import { setupDatumViews } from "../views/setupDatumViews";
import { connectDb } from "../auth/connectDb";
import { Show } from "../output/output";

export const command = "setup";
export const desc = "setup the database for use with datum";

export type SetupCmdArgs = BaseDatumArgs & {
  projectDir?: string;
};

export function builder(yargs: Argv): Argv {
  return yargs.options({
    "project-dir": {
      describe:
        "where to look for additional DatumViews to setup in the database",
      type: "string",
    },
  });
}

export async function setupCmd(args: SetupCmdArgs): Promise<void> {
  const db = connectDb(args);
  const show: Show = args.showAll ? Show.All : args.show ?? Show.None;
  await setupDatumViews({ db, show: show, projectDir: args.projectDir });
}
