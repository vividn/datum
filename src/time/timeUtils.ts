import { DateTime, Duration, Zone, Settings as DateTimeSettings } from "luxon";
import { BadDurationError, BadTimeError } from "../errors";
import { GenericObject } from "../GenericObject";
import { getTimezone } from "./getTimezone";

export type isoDatetime = string;
export type isoDate = string;
export type isoDateOrTime = isoDate | isoDatetime;
export type isoDuration = string;

type IANATimeZone = string;
export type DatumTime = {
  utc: isoDateOrTime;
  o?: number;
  tz?: IANATimeZone;
};

export function isIsoDateOrTime(str: string): str is isoDateOrTime {
  return DateTime.fromISO(str).isValid;
}

export function isDatumTime(
  time: DatumTime | GenericObject | string,
): time is DatumTime {
  if (typeof time === "string") {
    return false;
  }
  return (
    typeof time.utc === "string" &&
    isIsoDateOrTime(time.utc) &&
    (time.o === undefined || typeof time.o === "number") &&
    (time.tz === undefined || typeof time.tz === "string")
  );
}

export const now = (zone?: Zone | string): DateTime => DateTime.local({ zone });
export const defaultZone = DateTimeSettings.defaultZone as Zone;

export function isoDateFromDateTime(dt: DateTime): isoDate {
  return dt.toISODate() as isoDate;
}

export function isoDatetimeFromDateTime(dt: DateTime): isoDatetime {
  return dt.toUTC().toString() as isoDatetime;
}

export function isoDurationFromDuration(dur: Duration): isoDuration {
  if (dur.valueOf() < 0) {
    return "-" + dur.negate().toISO();
  }
  if (!dur.isValid) {
    throw new BadDurationError("invalid duration");
  }
  return dur.toISO() as isoDuration;
}

export function utcOffset(time: DateTime): number {
  const offset = time.offset / 60;
  // luxon sometimes gives -0 for the offset if system timezone is utc, so make it positive if necessary
  return Object.is(offset, -0) ? 0 : offset;
}

export function toDatumTime(
  time: DateTime | string,
  onlyDate?: boolean,
): DatumTime {
  if (typeof time === "string") {
    time = DateTime.fromISO(time);
  }
  // Checking if DateTime is valid should be done before calling this function
  if (!time.isValid) {
    throw new BadTimeError("invalid time was given");
  }
  if (onlyDate) {
    return { utc: isoDateFromDateTime(time) };
  }
  return {
    utc: isoDatetimeFromDateTime(time),
    o: utcOffset(time),
    tz: time.zone.name,
  };
}

export function datumTimeToLuxon(time?: DatumTime): DateTime | undefined {
  // returns undefined if utc is just a date or some invalid string
  if (
    time === undefined ||
    !isIsoDateOrTime(time.utc) ||
    !time.utc.includes("T")
  ) {
    return undefined;
  }
  if (time.tz && time.o) {
    const tzTime = DateTime.fromISO(time.utc, { zone: getTimezone(time.tz) });
    const oTime = DateTime.fromISO(time.utc, { zone: getTimezone(time.o) });
    if (tzTime.offset !== oTime.offset) {
      console.warn(
        `mismatched IANA timezone and offset for ${time.utc}. zone = ${time.tz}, o = ${time.o}`,
      );
      return oTime;
    }
  }
  const timezone = time.tz
    ? getTimezone(time.tz)
    : time.o
      ? getTimezone(time.o)
      : undefined;
  const dateTime = DateTime.fromISO(time.utc, {
    zone: timezone,
  });
  return dateTime.isValid ? dateTime : undefined;
}
