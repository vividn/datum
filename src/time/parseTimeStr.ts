import { DateTime, Duration } from "luxon";
import { now } from "./timeUtils";
import * as chrono from "chrono-node";
import { BadTimeError } from "../errors";
import { parseDurationStr } from "./parseDurationStr";

type ParseTimeStrType = {
  timeStr: string;
  referenceTime?: DateTime;
};

// TODO: Change this not to use an object input
export function parseTimeStr({
  timeStr,
  referenceTime,
}: ParseTimeStrType): DateTime {
  referenceTime = referenceTime ?? now();

  // Quick relative time strings (q = 5 minutes, t = 1 minute)
  const quickMatches = timeStr.match(/^([+-]?)([qt]+)$/i);
  if (quickMatches) {
    const sign = quickMatches[1] === "+" ? 1 : -1;
    const qs = quickMatches[2].match(/q/gi)?.length ?? 0;
    const ts = quickMatches[2].match(/t/gi)?.length ?? 0;
    const minutes = sign * (qs * 5 + ts);
    const duration = Duration.fromObject({ minutes });
    return referenceTime.plus(duration);
  }

  // This custom regex is to match a few extra strings not recognized by chrono, particularly short
  // E.g, 10 for 10:00, 1513 for 15:13, etc.
  const matches = timeStr.match(
    /^(?<hour>\d{1,2})(:?(?<minute>\d{2})(:?(?<second>\d{2})(\.(?<milli>\d{1,3}))?)?)?(?<meridian>am?|pm?)?$/i,
  );
  if (matches?.groups) {
    const {
      hour = "0",
      minute = "0",
      second = "0",
      milli = "0",
      meridian,
    } = matches.groups;

    // fairly dirty implementation, but built for speed
    const correctedHour =
      meridian?.toLowerCase() === "pm"
        ? parseInt(hour, 10) + 12
        : parseInt(hour, 10);

    // if less than all three millisecond digits are given, fill the remaining zeroes
    const millisecond = parseInt((milli + "000").substring(0, 3), 10);

    return referenceTime.set({
      hour: correctedHour,
      minute: parseInt(minute, 10),
      second: parseInt(second, 10),
      millisecond,
    });
  }

  // Also supports relative time strings, e.g., -5min
  if (timeStr.match(/^[+-]/)) {
    const duration = parseDurationStr({ durationStr: timeStr }) as Duration;
    return referenceTime.plus(duration);
  }

  // DateTime can parse some extra ISO type strings
  const dateTimeParsed = DateTime.fromISO(timeStr, { setZone: true });
  if (dateTimeParsed.isValid) {
    return dateTimeParsed;
  }

  // As a last resort, use chrono to parse the time
  const chronoParsed = chrono.parseDate(timeStr, {
    instant: referenceTime.toJSDate(),
    timezone: referenceTime.zone.offset(referenceTime.toMillis()),
  });
  if (chronoParsed) {
    return DateTime.fromISO(chronoParsed.toISOString(), {
      zone: referenceTime.zone,
    });
  }

  throw new BadTimeError(timeStr);
}
