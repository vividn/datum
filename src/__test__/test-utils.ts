import { CouchDbError } from "../errors";
import {
  DataOnlyDocument,
  DatumData,
  DatumDocument,
  DatumMetadata,
  EitherDocument,
  EitherPayload,
} from "../documentControl/DatumDocument";
import Mock = jest.Mock;
import { DateTime, Settings } from "luxon";
import { parseTimeStr } from "../time/parseTimeStr";
import { now } from "../time/timeUtils";
import { connectDb } from "../auth/connectDb";
import * as connectDbModule from "../auth/connectDb";
import { defaultIdComponents } from "../ids/defaultIdComponents";
import { buildIdStructure } from "../ids/buildIdStructure";
import { defaults } from "../input/defaults";
import { assembleId } from "../ids/assembleId";
import * as newHumanIdModule from "../meta/newHumanId";

// eslint-disable-next-line @typescript-eslint/no-empty-function
export const pass = (): void => {};
export const fail = (): never => {
  throw Error;
};

export const mockDocMissingError: CouchDbError = {
  status: 404,
  name: "not_found",
  error: "not_found",
  reason: "missing",
  message: "missing",
};

export const mockDocDeletedError: CouchDbError = {
  status: 404,
  message: "deleted",
  name: "not_found",
  error: "not_found",
  reason: "deleted",
};

export const mockMissingNamedViewError: CouchDbError = {
  status: 404,
  name: "not_found",
  error: "not_found",
  reason: "missing_named_view",
  message: "missing_named_view",
};

export async function resetTestDb(
  db: PouchDB.Database & {
    __opts?: PouchDB.Configuration.DatabaseConfiguration;
    _destroyed?: boolean;
    _closed?: boolean;
  },
): Promise<PouchDB.Database> {
  await db.destroy().catch(pass);
  // nasty hack to reopen closed database
  delete db._destroyed;
  delete db._closed;
  return db.constructor(db.name, db.__opts);
}

export function testDbLifecycle(
  dbName: string,
): PouchDB.Database<EitherPayload> {
  const db = connectDb({ db: dbName });

  beforeEach(async () => {
    await resetTestDb(db);
    jest.spyOn(connectDbModule, "connectDb").mockReturnValue(db);
  });

  afterEach(async () => {
    await db.destroy().catch(pass);
  });

  return db;
}

export function mockedLogLifecycle(): Mock {
  const originalLog = console.log;
  const mockedLog = jest.fn() as Mock;

  beforeEach(async () => {
    console.log = mockedLog;
  });

  afterEach(async () => {
    console.log = originalLog;
    mockedLog.mockReset();
  });

  return mockedLog;
}

const originalNowFn = Settings.now;
const nowStack: (() => number)[] = [];
export function setNow(timeStr: string): DateTime {
  const parsedTime = parseTimeStr({ timeStr, referenceTime: DateTime.now() });
  const newNow = () => parsedTime.toMillis();
  Settings.now = newNow;
  return parsedTime;
}

export function pushNow(timeStr: string): DateTime {
  nowStack.push(Settings.now);
  return setNow(timeStr);
}

export function popNow(): DateTime {
  const lastNow = nowStack.pop();
  if (lastNow === undefined) {
    throw new Error("tried to pop non existent now");
  }
  Settings.now = lastNow;
  return now();
}

export function restoreNow(): DateTime {
  Settings.now = originalNowFn;
  nowStack.length = 0;
  return now();
}

export function at<A extends any[], O>(
  timeStr: string,
  fn: (...args: A) => Promise<O>,
): (...args: A) => Promise<O> {
  return async (...args: A): Promise<O> => {
    pushNow(timeStr);
    const returnVal = await fn(...args);
    popNow();
    return returnVal;
  };
}

// TODO: Transition all tests to use makeDoc where appropriate
export function makeDoc(
  data: DatumData,
  meta: DatumMetadata | false = {},
  include_rev = false,
): EitherDocument {
  let doc: EitherDocument;
  if (meta === false) {
    doc = { ...data } as DataOnlyDocument;
  } else {
    doc = { ...data, meta } as DatumDocument;
  }

  meta = meta || {};
  if (doc._id === undefined) {
    let idStructure: string;
    if (meta?.idStructure) {
      idStructure = meta.idStructure;
    } else {
      const { defaultIdParts, defaultPartitionParts } = defaultIdComponents({
        data,
      });
      idStructure = buildIdStructure({
        idParts: defaultIdParts,
        delimiter: defaults.idDelimiter,
        partition: defaultPartitionParts,
      });
    }
    doc._id = assembleId({ payload: doc, idStructure });
  }

  if (include_rev) {
    doc._rev = "1-foo";
  }
  return doc;
}

export async function deterministicHumanIds(seed?: number): Promise<void> {
  let a = seed || 20231018;
  function random() {
    const x = Math.sin(a++) * 10000;
    return x - Math.floor(x);
  }
  function mockNewHumanId(): string {
    return random().toString(36).slice(2) + random().toString(36).slice(2);
  }

  beforeEach(() => {
    a = seed || 20231018;
    jest
      .spyOn(newHumanIdModule, "newHumanId")
      .mockImplementation(mockNewHumanId);
  });
}

// export async function generateSampleDay(dateStr = "2022-08-14") {
//
//   setNow(dateStr);
//   at('8:30', addCmd)({field: "sleep", })
//
//   popNow();
// }
