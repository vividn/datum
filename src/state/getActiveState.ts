import { DateTime } from "luxon";
import { activeStateView } from "../views/datumViews/activeStateView";
import { DatumTime, isDatumTime, now } from "../time/timeUtils";
import { DatumViewMissingError, isCouchDbError } from "../errors";
import { parseTimeStr } from "../time/parseTimeStr";
import { DatumState } from "./normalizeState";

export async function getActiveState(
  db: PouchDB.Database,
  field: string,
  time: DateTime | DatumTime | string = now(),
): Promise<DatumState> {
  const utcTime =
    typeof time === "string"
      ? parseTimeStr({ timeStr: time }).toUTC().toISO()
      : isDatumTime(time)
        ? time.utc
        : time.toUTC().toISO();
  if (utcTime === null) {
    throw new Error("bad time");
  }
  let viewResult;
  try {
    viewResult = await db.query(activeStateView.name, {
      startkey: [field, utcTime],
      descending: true,
      limit: 1,
    });
  } catch (error) {
    //TODO make special error checking missingDatumView
    if (
      isCouchDbError(error) &&
      ["missing", "deleted", "missing_named_view"].includes(error.reason)
    ) {
      throw new DatumViewMissingError(activeStateView.name);
    }
    throw error;
  }

  if (viewResult.rows.length === 0) {
    return null;
  }
  const activeStateRow = viewResult.rows[0];
  if (activeStateRow.key[0] !== field) {
    return null;
  }
  return activeStateRow.value;
}
