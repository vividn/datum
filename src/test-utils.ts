import Nano, { DocumentScope } from "nano";
import { CouchDbError } from "./errors";
import { EitherPayload } from "./documentControl/DatumDocument";
import * as connectDb from "./auth/connectDb";
import Mock = jest.Mock;
import { DateTime, Settings } from "luxon";
import { parseTimeStr } from "./time/parseTimeStr";
import { now } from "./time/timeUtils";

// eslint-disable-next-line @typescript-eslint/no-empty-function
export const pass = (): void => {};
export const fail = (): never => {
  throw Error;
};

export const testNano = Nano(`http://admin:password@localhost:5983`);

export const mockDocMissingError: CouchDbError = {
  scope: "couch",
  statusCode: 404,
  errid: "non_200",
  description: "missing",
  error: "not_found",
  reason: "missing",
};

export const mockDocDeletedError: CouchDbError = {
  scope: "couch",
  statusCode: 404,
  errid: "non_200",
  description: "deleted",
  error: "not_found",
  reason: "deleted",
};

export const mockMissingNamedViewError: CouchDbError = {
  scope: "couch",
  statusCode: 404,
  errid: "non_200",
  description: "missing_named_view",
  error: "not_found",
  reason: "missing_named_view",
};

export async function resetTestDb(
  dbName: string
): Promise<DocumentScope<EitherPayload>> {
  const maxTries = 3;
  let tries = 0;
  await testNano.db.destroy(dbName).catch(pass);
  while (tries++ <= maxTries) {
    await testNano.db.destroy(dbName).catch(pass);
    await testNano.db.create(dbName).catch(pass);
    if ((await testNano.db.list()).includes(dbName)) {
      return testNano.use(dbName) as DocumentScope<EitherPayload>;
    }
  }
  throw Error(`Unable to reset database after ${maxTries} attempts`);
}

export function testDbLifecycle(dbName: string): DocumentScope<EitherPayload> {
  const db = testNano.use(dbName) as DocumentScope<EitherPayload>;

  beforeEach(async () => {
    await resetTestDb(dbName);
    jest.spyOn(connectDb, "connectDb").mockImplementation(async () => db);
  });

  afterEach(async () => {
    await testNano.db.destroy(dbName).catch(pass);
  });

  return db;
}

export function mockedLogLifecycle(): Mock {
  const originalLog = console.log;
  const mockedLog = jest.fn() as Mock;

  beforeEach(async () => {
    console.log = mockedLog;
  });

  afterEach(async () => {
    console.log = originalLog;
    mockedLog.mockReset();
  });

  return mockedLog;
}

const originalNowFn = Settings.now;
const nowStack: (() => number)[] = [];
export function setNow(timeStr: string): DateTime {
  const parsedTime = parseTimeStr({ timeStr });
  const newNow = () => parsedTime.toMillis();
  Settings.now = newNow;
  return parsedTime;
}

export function pushNow(timeStr: string): DateTime {
  nowStack.push(Settings.now);
  return setNow(timeStr);
}

export function popNow(): DateTime {
  const lastNow = nowStack.pop();
  if (lastNow === undefined) {
    throw new Error("tried to pop non existent now");
  }
  Settings.now = lastNow;
  return now();
}

export function restoreNow(): DateTime {
  Settings.now = originalNowFn;
  nowStack.length = 0;
  return now();
}

export function at<A extends any[], O>(
  timeStr: string,
  fn: (...args: A) => O
): (...args: A) => O {
  return (...args: A): O => {
    pushNow(timeStr);
    const returnVal: O = fn(...args);
    popNow();
    return returnVal;
  };
}

// export async function generateSampleDay(dateStr = "2022-08-14") {
//
//   setNow(dateStr);
//   at('8:30', addCmd)({field: "sleep", })
//
//   popNow();
// }
