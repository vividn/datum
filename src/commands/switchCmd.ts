import { occurCmd, OccurCmdArgs } from "./occurCmd";
import { EitherDocument } from "../documentControl/DatumDocument";
import { flexiblePositional } from "../input/flexiblePositional";
import { durationArgs, DurationArgs } from "../input/durationArgs";

import { DatumState, normalizeState } from "../state/normalizeState";
import { ArgumentParser } from "argparse";
import { parseIfNeeded } from "../utils/parseIfNeeded";
import { dataArgs } from "../input/dataArgs";
import { timeArgs } from "../input/timeArgs";
import { fieldArgs } from "../input/fieldArgs";
import { dbArgs } from "../input/dbArgs";
import { outputArgs } from "../input/outputArgs";
import { newDocArgs } from "./addCmd";
import { inferType } from "../utils/inferType";

export const stateArgs = new ArgumentParser({
  add_help: false,
});
stateArgs.add_argument("state", {
  help: "the state to switch to, it defaults to true--equivalent to start",
  nargs: "?",
  default: "true",
});
stateArgs.add_argument("--last-state", "-L", {
  help: "manually specify the last state being transitioned out of",
  dest: "lastState",
});

export const switchArgs = new ArgumentParser({
  add_help: false,
  parents: [fieldArgs, stateArgs, durationArgs, newDocArgs, timeArgs, dataArgs],
});

export const switchCmdArgs = new ArgumentParser({
  description: "switch states of a given field",
  prog: "datum switch",
  usage: `%(prog)s <field> <state> [<duration OR .> [data..]]
  %(prog)s --moment <field> <state> [data..]
  %(prog)s <field> -k <reqKey> -k <optKey>=[defaultValue] ... <state> <duration OR .> <reqValue> [optValue] [data..]
`,
  parents: [switchArgs, dbArgs, outputArgs],
});

export type StateArgs = {
  state?: DatumState | string;
  lastState?: DatumState | string;
};
export type SwitchCmdArgs = OccurCmdArgs & DurationArgs & StateArgs;

export async function switchCmd(
  args: SwitchCmdArgs | string | string[],
  preparsed?: Partial<SwitchCmdArgs>,
): Promise<EitherDocument> {
  args = parseIfNeeded(switchCmdArgs, args, preparsed);
  flexiblePositional(
    args,
    "duration",
    "dur=",
    args.moment || args.omitTimestamp,
  );
  flexiblePositional(args, "state", "state=true");
  args.cmdData ??= {};
  if (args.moment) {
    args.cmdData.dur = null;
  }
  if (args.lastState) {
    const lastState =
      typeof args.lastState === "string"
        ? inferType(args.lastState)
        : args.lastState;
    if (lastState !== undefined) {
      args.cmdData.lastState = normalizeState(lastState);
    }
  }

  return await occurCmd(args);
}
