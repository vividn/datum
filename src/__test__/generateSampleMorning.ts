import { endCmd } from "../commands/endCmd";
import { occurCmd } from "../commands/occurCmd";
import { startCmd } from "../commands/startCmd";
import { switchCmd } from "../commands/switchCmd";
import { popNow, pushNow, setNow } from "./test-utils";

export async function generateSampleMorning(date: string): Promise<void> {
  pushNow(`8:30 ${date}`);
  await endCmd("sleep");
  setNow("+5");
  await startCmd("sleep -y -t 2320");
  setNow("8:40");
  await switchCmd("project german");
  await switchCmd("text fiction_book");
  setNow("9:10");
  await endCmd("project");
  await endCmd("text");
  setNow("9:30");
  await switchCmd("environment outside");
  await startCmd("stretch");
  setNow("+7");
  await endCmd("stretch");
  await startCmd("run");
  setNow("+30");
  await endCmd("run distance=5.4");
  await startCmd("stretch");
  setNow("+8");
  await endCmd("stretch");
  await switchCmd("environment home");
  setNow("+3");
  await occurCmd("pushup amount=10");
  setNow("10:30");
  await switchCmd("environment outside 5");
  setNow("11");
  await occurCmd("caffeine amount=100 -c coffee");
  popNow();
}
