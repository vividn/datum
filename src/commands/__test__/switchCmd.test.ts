import { restoreNow, setNow, testDbLifecycle } from "../../__test__/test-utils";
import { switchCmd } from "../switchCmd";
import { DateTime } from "luxon";
import { setupCmd } from "../setupCmd";
import { toDatumTime } from "../../time/timeUtils";
import { getActiveState } from "../../state/getActiveState";
import { parseTimeStr } from "../../time/parseTimeStr";

describe("switchCmd", () => {
  const dbName = "switch_cmd_test";
  const db = testDbLifecycle(dbName);

  beforeEach(async () => {
    await setupCmd({});
    setNow("2023-09-02,9:20");
  });
  afterEach(async () => {
    restoreNow();
  });

  it("creates an occur document with a custom state", async () => {
    const doc = await switchCmd({ field: "field", state: "newState" });
    expect(doc.data).toMatchObject({
      field: "field",
      state: "newState",
      occurTime: toDatumTime(DateTime.now()),
    });
    const dbDoc = await db.get(doc._id);
    expect(dbDoc).toEqual(doc);
  });

  it("records what it thinks the last state was", async () => {
    const doc = await switchCmd({ field: "environment", state: "outside" });
    expect(doc.data).toMatchObject({
      field: "environment",
      state: "outside",
      lastState: null,
    });
    setNow("+5");
    const doc2 = await switchCmd({ field: "environment", state: "inside" });
    expect(doc2.data).toMatchObject({
      field: "environment",
      state: "inside",
      lastState: "outside",
    });
  });

  it("can take in a manually specified last state", async () => {
    await switchCmd({ field: "project", state: "datum" });
    setNow("+5");
    const doc = await switchCmd({
      field: "project",
      state: "household",
      lastState: "german",
    });
    expect(doc.data).toMatchObject({
      field: "project",
      state: "household",
      lastState: "german",
    });
  });

  it("does not include a lastState if there is no occurTime", async () => {
    await switchCmd({ field: "vacuum", state: true });
    const doc = await switchCmd({
      field: "vacuum",
      state: false,
      noTimestamp: true,
    });
    expect(doc.data).toMatchObject({
      field: "vacuum",
      state: false,
    });
    expect(doc.data).not.toHaveProperty("lastState");
    expect(doc.data).not.toHaveProperty("occurTime");
  });

  it("still includes lastState if manually specified even when there is no occurTime", async () => {
    await switchCmd({ field: "project", state: "linux" });
    const doc = await switchCmd({
      field: "project",
      state: "backend",
      lastState: "frontend",
      noTimestamp: true,
    });
    expect(doc.data).toMatchObject({
      field: "project",
      state: "backend",
      lastState: "frontend",
    });
    expect(doc.data).not.toHaveProperty("occurTime");
  });

  it("can do a state of null", async () => {
    const doc = await switchCmd({ field: "vacuum", state: null });
    const doc2 = await switchCmd({ field: "luft", state: "null" });
    expect(doc.data).toMatchObject({
      field: "vacuum",
      state: null,
    });
    expect(doc2.data).toMatchObject({
      field: "luft",
      state: null,
    });
  });

  it.todo(
    "assumes a state of true by default",
    // TODO: Write this once commands can be called by plain text in tests
  );

  it("records a block of state when specifying duration", async () => {
    const doc = await switchCmd({
      field: "project",
      state: "household",
      duration: "10m",
    });
    expect(doc.data).toMatchObject({
      field: "project",
      state: "household",
      dur: "PT10M",
    });
    const intermediateState = await getActiveState(
      db,
      "project",
      parseTimeStr({ timeStr: "-5m" }),
    );
    expect(intermediateState).toBe("household");
  });

  it("records a hole of time in the state when given a negative duration", async () => {
    const doc = await switchCmd({
      field: "project",
      state: "household",
      duration: "-10m",
    });
    expect(doc.data).toMatchObject({
      field: "project",
      state: "household",
      dur: "-PT10M",
    });
    const intermediateState = await getActiveState(
      db,
      "project",
      parseTimeStr({ timeStr: "-5m" }),
    );
    expect(intermediateState).toBe(false);
  });

  it("when --moment is specified, dur is null and there is no duration postional argument", async () => {
    const doc = await switchCmd({
      field: "project",
      state: "household",
      moment: true,
      optional: ["skillPoints"],
      duration: "3",
    });
    expect(doc.data).toMatchObject({
      field: "project",
      state: "household",
      skillPoints: 3,
      dur: null,
      occurTime: toDatumTime(DateTime.now()),
    });
  });

  it("when --no-timestamp is specified, there is no positional duration argument", async () => {
    const doc = await switchCmd({
      field: "project",
      state: "household",
      optional: ["skillPoints"],
      noTimestamp: true,
      duration: "3",
    });
    expect(doc.data).toMatchObject({
      field: "project",
      state: "household",
      skillPoints: 3,
    });
    expect(doc.data).not.toHaveProperty("occurTime");
    expect(doc.data).not.toHaveProperty("dur");
  });

  it("can skip the duration if the duration is given as . or ''", async () => {
    // TODO: rewrite this test as a string based call;
    restoreNow();
    const doc = await switchCmd({
      field: "project",
      state: "household",
      moment: false,
      optional: "optional",
      duration: ".",
      data: [50],
    });
    const doc2 = await switchCmd({
      field: "project",
      state: "household",
      moment: false,
      optional: "optional",
      duration: "",
      data: [50],
    });
    expect(doc.data).toMatchObject({ field: "project", optional: 50 });
    expect(doc.data).not.toHaveProperty("dur");
    expect(doc2.data).toMatchObject({ field: "project", optional: 50 });
    expect(doc2.data).not.toHaveProperty("dur");
  });

  it.todo("does not record a lastState if there is no occurTime");
});
