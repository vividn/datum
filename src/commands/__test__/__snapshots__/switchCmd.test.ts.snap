// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`switchCmd change command can become a start command by having start as a trailing word 1`] = `
{
  "_id": "field:2023-12-21T14:00:00.000Z",
  "_rev": Any<String>,
  "data": {
    "dur": "PT5M30S",
    "field": "field",
    "key": "val",
    "lastState": null,
    "occurTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
    "opt1": "optVal",
    "state": "someState",
  },
  "meta": {
    "createTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
    "humanId": "elfuv1633zeysuyhc0l9ss",
    "idStructure": "%field%:%occurTime%",
    "modifyTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
  },
}
`;

exports[`switchCmd change command can become an end command by having start as a trailing word 1`] = `
{
  "_id": "field:2023-12-21T14:00:00.000Z",
  "_rev": Any<String>,
  "data": {
    "dur": "PT30M",
    "field": "field",
    "key": "val",
    "lastState": "someState",
    "occurTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
    "opt1": "optVal",
    "state": false,
  },
  "meta": {
    "createTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
    "humanId": "elfuv1633zeysuyhc0l9ss",
    "idStructure": "%field%:%occurTime%",
    "modifyTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
  },
}
`;

exports[`switchCmd change command can become an occur command by having occur as a trailing word 1`] = `
{
  "_id": "field:2023-12-21T14:00:00.000Z",
  "_rev": Any<String>,
  "data": {
    "dur": null,
    "field": "field",
    "key": "val",
    "lastState": null,
    "occurTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
    "opt1": "optVal",
    "state": "someState",
  },
  "meta": {
    "createTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
    "humanId": "elfuv1633zeysuyhc0l9ss",
    "idStructure": "%field%:%occurTime%",
    "modifyTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
  },
}
`;
